client/dashboard/domains/name-servers/form.tsx

```diff
@@ -0,0 +1,219 @@
+import UpsellNudge from './upsell-nudge';
+import { areAllWpcomNameServers } from './utils';
+import type { InstanceType } from '../../app/context';
+
+interface Props {
+	domainName: string;
+	instanceType: InstanceType;
+	showUpsellNudge?: boolean;
+	nameservers?: string[];
+	isBusy?: boolean;
+	queryError?: string;
+	onSubmit: ( nameServers: string[] ) => void;
+}
+
+export default function NameServersForm( {
+	domainName,
+	instanceType,
+	showUpsellNudge,
+	nameservers = [],
+	isBusy,
+	queryError,
+	onSubmit,
+}: Props ) {
+	const [ nameServerFields, setNameServerFields ] = useState< NameServerField[] >(
+		Array.from( { length: MAX_NAMESERVER_LENGTH }, ( _, index ) => ( {
+			value: nameservers[ index ] || '',
+			error: '',
+			touched: false,
+		} ) )
+	);
+	const [ useWpcomNameservers, setUseWpcomNameservers ] = useState(
+		areAllWpcomNameServers( nameservers )
+	);
+
+	const hasFieldsChanged = useMemo( () => {
+		// Get current non-empty values
+		const currentValues = nameServerFields
+			.filter( ( ns ) => ns.value !== '' )
+			.map( ( ns ) => ns.value )
+			.sort();
+
+		// Get initial values
+		const initialValues = [ ...nameservers ].sort();
+
+		// Compare arrays
+		if ( currentValues.length !== initialValues.length ) {
+			return true;
+		}
+
+		return currentValues.some( ( value, index ) => value !== initialValues[ index ] );
+	}, [ nameServerFields, nameservers ] );
+
+	const isFormValid = useMemo( () => {
+		// Check if there are any errors
+		if ( nameServerFields.some( ( ns ) => ns.error !== '' ) ) {
+			return false;
+		}
+
+		// Check if minimum required nameservers are provided
+		const filledNameServers = nameServerFields.filter( ( ns ) => ns.value !== '' ).length;
+		return filledNameServers >= MIN_NAMESERVER_LENGTH;
+	}, [ nameServerFields ] );
+
+	const canSubmit = useMemo( () => {
+		return isFormValid && hasFieldsChanged && ! isBusy;
+	}, [ isFormValid, hasFieldsChanged, isBusy ] );
+
+	const handleNameServerChange = useCallback( ( index: number, value: string ) => {
+		setNameServerFields( ( current ) => {
+			const updated = [ ...current ];
+			updated[ index ] = {
+				...updated[ index ],
+				value,
+				error: updated[ index ].touched ? validateField( value, index ) : '',
+			};
+			return updated;
+		} );
+	}, [] );
+
+	const handleNameServerBlur = useCallback( ( index: number ) => {
+		setNameServerFields( ( current ) => {
+			if ( current[ index ].touched ) {
+				return current;
+			}
+			const updated = [ ...current ];
+			updated[ index ] = {
+				...updated[ index ],
+				touched: true,
+				error: validateField( updated[ index ].value, index ),
+			};
+			return updated;
+		} );
+	}, [] );
+
+	const shouldShowNextInput = useCallback(
+		( index: number ) => {
+			// When using WP.com nameservers, only show fields that have values
+			if ( useWpcomNameservers ) {
+				return nameServerFields[ index ]?.value !== '';
+			}
+
+			// For custom nameservers
+			if ( index >= MAX_NAMESERVER_LENGTH ) {
```

Comment:
This feels similar to the validation function we have in the form input. Can't we reuse it here?